internal/wazeroir/operations.go

```diff
@@ -2140,336 +2068,205 @@ const (
-
-// Kind implements Operation.Kind.
-func (OperationV128ExtractLane) Kind() OperationKind {
-	return OperationKindV128ExtractLane
+//
+// laneIndex is >=0 && <M where shape = NxM.
+// signed is used when shape is either i8x16 or i16x2 to specify whether to sign-extend or not.
+func NewOperationV128ExtractLane(laneIndex byte, signed bool, shape Shape) UnionOperation {
+	return UnionOperation{
+		OpKind: OperationKindV128ExtractLane,
+		B1:     shape,
+		B2:     laneIndex,
+		B3:     signed,
+	}
 }
 
-// OperationV128ReplaceLane implements Operation.
+// NewOperationV128ReplaceLane is a constructor for UnionOperation with Kind OperationKindV128ReplaceLane.
 //
 // This corresponds to
 //
 //	wasm.OpcodeVecI8x16ReplaceLaneName wasm.OpcodeVecI16x8ReplaceLaneName
 //	wasm.OpcodeVecI32x4ReplaceLaneName wasm.OpcodeVecI64x2ReplaceLaneName
 //	wasm.OpcodeVecF32x4ReplaceLaneName wasm.OpcodeVecF64x2ReplaceLaneName.
-type OperationV128ReplaceLane struct {
-	// LaneIndex is >=0 && <M where shape = NxM.
-	LaneIndex byte
-	Shape     Shape
-}
-
-// String implements fmt.Stringer.
-func (o OperationV128ReplaceLane) String() string { return o.Kind().String() }
-
-// Kind implements Operation.Kind.
-func (OperationV128ReplaceLane) Kind() OperationKind {
-	return OperationKindV128ReplaceLane
+//
+// laneIndex is >=0 && <M where shape = NxM.
+func NewOperationV128ReplaceLane(laneIndex byte, shape Shape) UnionOperation {
+	return UnionOperation{OpKind: OperationKindV128ReplaceLane, B1: shape, B2: laneIndex}
 }
 
-// OperationV128Splat implements Operation.
+// NewOperationV128Splat is a constructor for UnionOperation with Kind OperationKindV128Splat.
 //
 // This corresponds to
 //
 //	wasm.OpcodeVecI8x16SplatName wasm.OpcodeVecI16x8SplatName
 //	wasm.OpcodeVecI32x4SplatName wasm.OpcodeVecI64x2SplatName
 //	wasm.OpcodeVecF32x4SplatName wasm.OpcodeVecF64x2SplatName.
-type OperationV128Splat struct {
-	Shape Shape
-}
-
-// String implements fmt.Stringer.
-func (o OperationV128Splat) String() string { return o.Kind().String() }
-
-// Kind implements Operation.Kind.
-func (OperationV128Splat) Kind() OperationKind {
-	return OperationKindV128Splat
+func NewOperationV128Splat(shape Shape) UnionOperation {
+	return UnionOperation{OpKind: OperationKindV128Splat, B1: shape}
 }
 
-// OperationV128Shuffle implements Operation.
-type OperationV128Shuffle struct {
-	Lanes [16]byte
-}
-
-// String implements fmt.Stringer.
-func (o OperationV128Shuffle) String() string { return o.Kind().String() }
-
-// Kind implements Operation.Kind.
-//
-// This corresponds to wasm.OpcodeVecV128i8x16ShuffleName.
-func (OperationV128Shuffle) Kind() OperationKind {
-	return OperationKindV128Shuffle
+// NewOperationV128Shuffle is a constructor for UnionOperation with Kind OperationKindV128Shuffle.
+func NewOperationV128Shuffle(lanes [16]byte) UnionOperation {
+	op := UnionOperation{OpKind: OperationKindV128Shuffle}
+	op.Us = make([]uint64, 16)
```

Comment:
Unnecessary copy and allocation here. Let's just make this function accept []uint64 in the first place.
```suggestion
func NewOperationV128Shuffle(lanes []uint64) UnionOperation {
	op := UnionOperation{OpKind: OperationKindV128Shuffle, Us: lanes}
```